include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -0,0 +1,616 @@
+            cuvm::Unmap(block_ptr(i), block_size);
+            CUDA_DTOR_CALL(cuMemRelease(mapping[i]));
+            mapping[i] = {};
+            if (available[i])
+              no_longer_in_range++;
+          }
+        }
+        available_blocks -= no_longer_in_range;
+      }
+      mapping.resize(new_num_blocks);
+      mapped.resize(new_num_blocks);
+      available.resize(new_num_blocks);
+      address_range.size() = new_num_blocks * block_size;
+    }
+
+    void set_block_availability(char *start, char *end, bool availability) {
+      int blk_start = (start - block_ptr<char>(0)) / block_size;
+      int blk_end   = (end   - block_ptr<char>(0)) / block_size;
+      int flipped = 0;
+      for (int b = available.find(!availability, blk_start); b < blk_end; ) {
+        int first_flipped = available.find(availability, b+1);
+        flipped += first_flipped - b;
+        b = first_flipped + 1;
+      }
+      if (availability)
+        available_blocks += flipped;
+      else
+        available_blocks -= flipped;
+      available.fill(blk_start, blk_end, availability);
+    }
+
+    cuvm::CUAddressRange    address_range;
+    vector<mem_handle_t>    mapping;
+    bitmask                 mapped, available;
+    ptrdiff_t               available_blocks = 0;
+    size_t                  block_size = 0;
+  };
+
+  SmallVector<va_region, 8> va_regions_;
+  SmallVector<cuvm::CUMemAddressRange, 8> va_ranges_;
+  size_t initial_va_size_ = 0;
+  size_t block_size_ = 0;
+  size_t total_mem_ = 0;
+  int device_ordinal_ = -1;
+
+  void adjust_params(size_t &size, size_t &alignment) {
+    alignment = std::max(alignment, next_pow2(size >> 11));
+    alignment = std::min(alignment, block_size_);
+    size = align_up(size, alignment);
+  }
+
+  void *do_allocate(size_t size, size_t alignment) override {
+    adjust_params(size, alignment);
+    if (size == 0)
+      return nullptr;
+    lock_guard pool_guard(pool_lock_);
+    // try to get a free region that's already mapped
+    void *ptr = try_get_mapped(size, alignment);
+    if (ptr) {
+      stat_add_allocation(size);
+      return ptr;
+    }
+    DeviceGuard dg(device_ordinal_);
+    mem_lock_guard mem_guard(mem_lock_);
+    ptr = get_va(size, alignment);
+    map_storage(ptr, size);
+    ptr = try_get_mapped(size, alignment);
+    assert(ptr != nullptr);
+    stat_add_allocation(size);
+    return ptr;
+  }
+
+  void *try_get_mapped(size_t size, size_t alignment) {
+    char *ptr = static_cast<char*>(free_mapped_.get(size, alignment));
+    if (ptr) {
+      stat_take_free(size);
+      free_va_.get_specific_block(ptr, ptr + size);
+      mark_as_unavailable(ptr, size);
+      return ptr;
+    } else {
+      return nullptr;
+    }
+  }
+
+  void *get_va(size_t size, size_t alignment) {
+    void *ptr = free_va_.get(size, alignment);
+    if (ptr)
+      return ptr;
+    va_allocate(size);
+    ptr = free_va_.get(size, alignment);
+    assert(ptr);
+    return ptr;
```

Comment:
What would happen if we return nullptr here? Would it properly error or just crash?

Reply:
It would crash. This function is executed under a lock_guard. The job of `va_allocate` is to make sure that we _can_ make the allocation of properly aligned `size` bytes from free_va. If that is impossible to achieve, then va_allocate should throw.

Reply:
Throw is good enogh.